composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,379 @@
+            self.index = StreamingDatasetIndex.load(fp)
+
+        # Fields, protected by the lock, relating to loading shards in the background.
+        self._lock = None
+        self._next_epoch = 0
+        self._epoch_to_todo_ids = {}
+        self._downloaded_ids = []
+        self._are_all_shards_downloaded = False
+
+    def _download_if_missing(self, basename: str) -> str:
+        """Safely download a shard from remote to local cache.
+
+        Args:
+            basename (str): Basename of shard to download.
+
+        Returns:
+            str: Local cache filename.
+        """
+        remote = os.path.join(self.remote, basename)
+        local = os.path.join(self.local, basename)
+        safe_download(remote, local)
+        return local
+
+    def _load_shards(self, shards: Sequence[int], part_min_id: int, part_max_id: int) -> None:
+        """Load the given list of locally cached shards into the dataset.
+
+        Every time you call __iter__ on this dataset, it registers the list of
+        samples you have left, which will not be the full epoch if the dataset
+        isn't finished loaded when you start training.
+
+        Calls to _load_shards during training modify the samples remaining on
+        these iterations on the fly to insert these new samples and then resort,
+        making the shuffle as perfect as was possible.
+
+        This operation takes the lock, so batch your _load_shards calls where
+        possible.
+
+        Args:
+            shards (Sequence[int]): List of shards to load.
+            part_min_id (int): Minimum sample ID of this partition.
+            part_max_id (int): Maximum sample ID of this partition.
+        """
+        # Get all samples from the given shards that fall within our partition.
+        new_ids = []
+        for shard in shards:
+            shard_min_id = self.index.shard_begins[shard]
+            shard_max_id = self.index.shard_ends[shard] - 1
+            min_id = max(part_min_id, shard_min_id)
+            max_id = min(part_max_id, shard_max_id)
+            new_ids += list(range(min_id, max_id + 1))
+
+        with self._lock:
+            # Extend and optionally reshuffle the remaining samples of any
+            # epochs we have in progress.
+            if self.shuffle:
+                if not self._are_all_shards_downloaded:
+                    self._downloaded_ids.extend(new_ids)
+                    np.random.shuffle(self._downloaded_ids)
+                for todo_ids in self._epoch_to_todo_ids.values():
+                    todo_ids.extend(new_ids)
+                    np.random.shuffle(todo_ids)
+            else:
+                if not self._are_all_shards_downloaded:
+                    self._downloaded_ids.extend(new_ids)
+                for todo_ids in self._epoch_to_todo_ids.values():
+                    todo_ids.extend(new_ids)
+
+    def _load_shards_if_downloaded(self, shards: Sequence[int], part_min_id: int, part_max_id: int) -> List[int]:
```

Comment:
For code readability, would it make sense to rename this function `determine_downloaded_and_missing_shards(self) -> Tuple[downloaded_shards, missing_shards]`, and then have `_load` process over the downloaded shards and missing shards as appropriate?